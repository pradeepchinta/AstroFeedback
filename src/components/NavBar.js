
import React from "react";
import { Navbar, Nav, Form, FormControl, Button } from "react-bootstrap";

const NavBar = ({ handleSearch, showSearch, searchText, handleChange, searchEnabled }) => {
    return (
        <Navbar bg="light" expand="lg">
            <Navbar.Brand href="/">Astro Customer Feedbacks</Navbar.Brand>
            <Navbar.Toggle aria-controls="basic-navbar-nav" />
            <Navbar.Collapse id="basic-navbar-nav">
                <Nav className="mr-auto">
                    <Nav.Link href="/add-new">Add New Feedback</Nav.Link>
                </Nav>
                {showSearch && <Form inline>
                    <FormControl type="text" placeholder="Search Using Email" className="mr-sm-2" value={searchText} onChange={handleChange}/>
                    <Button id="SearchButton" variant="outline-success" onClick={handleSearch} disabled={searchEnabled} >Search</Button>
                </Form>}
            </Navbar.Collapse>
        </Navbar>
    )
};

export default NavBar;