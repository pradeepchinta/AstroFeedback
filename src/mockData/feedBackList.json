[
    {
      "id": "1",
      "title": "title 1",
      "type": "type 1",
      "description": "description 1",
      "email": "email 1",
      "date": 1593590610,
      "like": 19,
      "dislike": 88
    },
    {
      "id": "2",
      "title": "title 2",
      "type": "type 2",
      "description": "description 2",
      "email": "email 2",
      "date": 1593590550,
      "like": 80,
      "dislike": 9
    },
    {
      "id": "3",
      "title": "title 3",
      "type": "type 3",
      "description": "description 3",
      "email": "email 3",
      "date": 1593590490,
      "like": 10,
      "dislike": 27
    },
    {
      "id": "4",
      "title": "title 4",
      "type": "type 4",
      "description": "description 4",
      "email": "email 4",
      "date": 1593590430,
      "like": 39,
      "dislike": 41
    }
]