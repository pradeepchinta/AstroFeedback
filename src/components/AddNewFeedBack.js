import React, { useState, useRef, useEffect } from "react";
import { Container, Alert, Form, Button } from "react-bootstrap";
import moment from 'moment';
import { connect } from 'react-redux';

import NavBar from "./NavBar";
import { saveFeedback, hideMessage, showMessage } from '../actions';

export const NewFeedBack = ({ saveFeedback, responseMessage, responseVariant, hideMessage, showMessage }) => {
    const [validated, setValidated] = useState(false);
    const formRef = useRef(null);

    useEffect(() => {
        setTimeout(() => {
            hideMessage();
            if (responseMessage === "FeedBack Created Sucessfully") {
                formRef.current.reset();
                setValidated(false);
            }
        }, 2000)
    }, [responseMessage])

    const renderAlert = () => {
        return (
            <Alert variant={responseVariant}>
                {responseMessage}
            </Alert>
        )
    };

    const handleSubmit = (event) => {
        event.preventDefault();
        const form = event.currentTarget;
        const reg = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
        if (form.checkValidity() === true) {
            const data = {
                title: form.formBasicTitle.value,
                type: form.formBasicType.value,
                description: form.formBasicDescription.value,
                email: form.formBasicEmail.value,
                date: moment(),
                like: 0,
                disLike: 0
            };
            if (reg.test(form.formBasicEmail.value)) {
                saveFeedback(data);
            } else {
                showMessage({
                    message: "Please enter valid email ...!",
                    variant: "danger"
                })
            }
        }
        setValidated(true);
    };



    return (
        <Container fluid>
            <NavBar showSearch={false} />
            <Form ref={formRef} noValidate validated={validated} onSubmit={handleSubmit}>
                <Form.Group controlId="formBasicTitle">
                    <Form.Label>Title</Form.Label>
                    <Form.Control required type="title" placeholder="Title" />
                    <Form.Control.Feedback type="invalid">Please Provide Feedback title</Form.Control.Feedback>
                </Form.Group>
                <Form.Group controlId="formBasicType">
                    <Form.Label>Type</Form.Label>
                    <Form.Control required as="select">
                        <option value="">-- Select --</option>
                        <option value="Technical">Technical</option>
                        <option value="User">User</option>
                        <option value="Management Related">Management Related</option>
                        <option value="competitor Related">competitor Related</option>
                        <option value="Pricing issue">Pricing issue</option>
                        <option value="Customer Care issue">Customer Care issue</option>
                    </Form.Control>
                    <Form.Control.Feedback type="invalid">Please Select Feedback type</Form.Control.Feedback>
                </Form.Group>
                <Form.Group controlId="formBasicEmail">
                    <Form.Label>Email address</Form.Label>
                    <Form.Control required type="email" placeholder="Enter email" onChange={hideMessage} />
                    <Form.Control.Feedback type="invalid">Please Enter valid Email</Form.Control.Feedback>
                </Form.Group>
                <Form.Group controlId="formBasicDescription">
                    <Form.Label>Description</Form.Label>
                    <Form.Control required as="textarea" rows="3" />
                    <Form.Control.Feedback type="invalid">Please Enter Feedback Description</Form.Control.Feedback>
                </Form.Group>
                {(responseMessage) && renderAlert()}
                <Button variant="primary" type="submit">
                    Submit
                </Button>
            </Form>
        </Container>
    )
};

const mapStateToProps = (state) => {
    return {
        feedBackList: state.feedBackReducer.feedBackList,
        responseMessage: state.feedBackReducer.responseMessage,
        responseVariant: state.feedBackReducer.responseVariant
    };
};
const mapDispatchToProps = (dispatch) => {
    return {
        saveFeedback: (data) => dispatch(saveFeedback(data)),
        hideMessage: () => dispatch(hideMessage()),
        showMessage: (data) => dispatch(showMessage(data))
    };
};
export default connect(mapStateToProps, mapDispatchToProps)(NewFeedBack);