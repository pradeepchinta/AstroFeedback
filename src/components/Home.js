import React, { useEffect, useState } from "react";
import { connect } from 'react-redux';
import moment from 'moment';
import { Container, Media, Card, Spinner, Alert, Button } from "react-bootstrap";

import NavBar from "./NavBar";
import { fetchFeedBacks, searchFeedbacks, hideMessage, showMessage, updateFeedback } from '../actions';
import personImage from '../assets/person.png';

export const Home = ({ feedBackList, fetchFeedBacks, searchFeedbacks, responseMessage, responseVariant, showMessage, hideMessage, updateFeedback }) => {

    const [searchText, setSearchText] = useState("");
    const [searchEnabled, setSearchEnabled] = useState(false);
    const [liked, setLiked] = useState(false);
    const [disliked, setDisliked] = useState(false);
    //const [like, setLike] = useState(0);
    //const [disLike, setDisLike] = useState(0);
    let like = 0;
    let disLike = 0;

    useEffect(() => {
        fetchFeedBacks();
    }, []);

    useEffect(() => {
        setTimeout(() => {
            hideMessage();
        }, 2000)
    }, [responseMessage])

    const handleSearch = () => {
        const reg = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
        if (!searchText) {
            fetchFeedBacks();
            setSearchEnabled(false);
            return
        }
        if (reg.test(searchText)) {
            searchText && searchFeedbacks(searchText);
            setSearchEnabled(true);
        } else {
            showMessage({
                message: "Please enter valid email to search ...!",
                variant: "danger"
            })
            setSearchEnabled(false);
        }
    };

    const handleChange = (e) => {
        setSearchText(e.target.value);
    }

    const handleClick = (e, item) => {
        switch (e.target.name) {
            case "like":
                if (liked) {
                    like = item.like -1;
                    disLike = item.disLike;
                    setLiked(false);
                } else {
                    like = ++item.like;
                    disLike = item.disLike;
                    setLiked(true);
                }
                break;
            case "dislike":
                if (disliked) {
                    disLike = item.disLike -1;
                    like = item.like;
                    setDisliked(false);
                } else {
                    disLike = ++item.disLike;
                    like = item.like;
                    setDisliked(true);
                }
                break;
            default:
                return null;
        };
        const data = {
            id: item.id,
            title: item.title,
            type: item.type,
            description: item.description,
            email: item.email,
            date: item.date,
            like: like,
            disLike: disLike
        };
        updateFeedback(data);
    }

    const compare = (item, item1) => {
        if (item1.like > item.like) return 1;
        if (item.like > item1.like) return -1;
        return 0;
    }

    const renderList = () => {
        return feedBackList && feedBackList.sort(compare).map((item) => {
            return <div key={item.id}>
                <Media as="li">
                    <img
                        width={64}
                        height={64}
                        className="mr-3"
                        src={personImage}
                        alt="Generic placeholder"
                    />
                    <Media.Body>
                        <h5>{item.title}</h5>
                        <strong>{item.type}</strong>
                        <p>{item.description}</p>
                        <p>{item.email}</p>
                        <p>{moment(item.date).format("DD-MM-YYYY")}</p>
                        <Button id="like" onClick={(e) => handleClick(e, item)} variant="outline-primary" type="button" name="like">{`Like ${item.like}`}</Button>{' '}
                        <Button id="dislike" onClick={(e) => handleClick(e, item)} variant="outline-primary" type="button" name="dislike">{`DisLike ${item.disLike}`}</Button>
                    </Media.Body>
                </Media>
                <hr />
            </div >
        })
    };

    const renderSpinner = () => {
        return (
            <React.Fragment>
                <Spinner animation="border" role="status">
                    <span className="sr-only">Loading...</span>
                </Spinner>
                {feedBackList && !!feedBackList.length && <h2>We have Empty FeedBack List ...</h2>}
            </React.Fragment>
        )
    };

    const renderAlert = () => {
        return (
            <Alert variant={responseVariant}>
                {responseMessage}
            </Alert>
        )
    };

    return (
        <Container fluid>
            <NavBar handleSearch={handleSearch} showSearch={true} handleChange={handleChange} searchText={searchText} searchEnabled={searchEnabled}/>
            {(responseMessage && responseVariant) && renderAlert()}
            <Card>
                {feedBackList && feedBackList.length ? renderList() : (
                    <div className='spinner'>
                        {searchEnabled ? <h2>No FeedBack List Found</h2> : renderSpinner()}
                    </div>)}
            </Card>
        </Container>
    )
};

const mapStateToProps = (state) => {
    return {
        feedBackList: state.feedBackReducer.feedBackList,
        responseMessage: state.feedBackReducer.responseMessage,
        responseVariant: state.feedBackReducer.responseVariant
    };
};
const mapDispatchToProps = (dispatch) => {
    return {
        fetchFeedBacks: () => dispatch(fetchFeedBacks()),
        searchFeedbacks: (textSearch) => dispatch(searchFeedbacks(textSearch)),
        updateFeedback: (data) => dispatch(updateFeedback(data)),
        hideMessage: () => dispatch(hideMessage()),
        showMessage: (data) => dispatch(showMessage(data))
    };
};
export default connect(mapStateToProps, mapDispatchToProps)(Home);