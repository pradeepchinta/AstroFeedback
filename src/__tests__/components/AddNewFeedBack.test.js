import React from "react";
import { mount } from "enzyme";
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { NewFeedBack } from "../../components/AddNewFeedBack";
import feedBackList from "../../mockData/feedBackList";

const middlewares = [thunk]
const mockStore = configureStore(middlewares);
const store = mockStore({});

describe("NewFeedBack Component", () => {
    let wrapper;
    const minProps = {
        feedBackList,
        responseMessage: "Created Successfully",
        responseVariant: "success",
        saveFeedback: jest.fn(),
        hideMessage: jest.fn(),
        showMessage: jest.fn()
    };

    it("will render with min props", () => {
        wrapper = mount(<NewFeedBack {...minProps} store={store} />);
        expect(wrapper).toMatchSnapshot();
    })
    it("will render NavBar Component", () => {
        expect(wrapper.find("NavBar").length).toBe(1);
    })
    it("will Form with fields", () => {
        expect(wrapper.find("Form").length).toBe(1);
        expect(wrapper.find("Form").children().length).toBe(1);
    })
    it("will render Alert Message", () => {
        expect(wrapper.find("Alert").props().children).toBe("Created Successfully");
    })
})