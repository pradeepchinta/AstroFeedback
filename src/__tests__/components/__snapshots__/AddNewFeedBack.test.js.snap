// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewFeedBack Component will render with min props 1`] = `
<NewFeedBack
  feedBackList={
    Array [
      Object {
        "date": 1593590610,
        "description": "description 1",
        "dislike": 88,
        "email": "email 1",
        "id": "1",
        "like": 19,
        "title": "title 1",
        "type": "type 1",
      },
      Object {
        "date": 1593590550,
        "description": "description 2",
        "dislike": 9,
        "email": "email 2",
        "id": "2",
        "like": 80,
        "title": "title 2",
        "type": "type 2",
      },
      Object {
        "date": 1593590490,
        "description": "description 3",
        "dislike": 27,
        "email": "email 3",
        "id": "3",
        "like": 10,
        "title": "title 3",
        "type": "type 3",
      },
      Object {
        "date": 1593590430,
        "description": "description 4",
        "dislike": 41,
        "email": "email 4",
        "id": "4",
        "like": 39,
        "title": "title 4",
        "type": "type 4",
      },
    ]
  }
  hideMessage={[MockFunction]}
  responseMessage="Created Successfully"
  responseVariant="success"
  saveFeedback={[MockFunction]}
  showMessage={[MockFunction]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Container
    fluid={true}
  >
    <div
      className="container-fluid"
    >
      <NavBar
        showSearch={false}
      >
        <Navbar
          bg="light"
          collapseOnSelect={false}
          expand="lg"
          variant="light"
        >
          <nav
            className="navbar navbar-expand-lg navbar-light bg-light"
          >
            <NavbarBrand
              href="/"
            >
              <a
                className="navbar-brand"
                href="/"
              >
                Astro Customer Feedbacks
              </a>
            </NavbarBrand>
            <NavbarToggle
              aria-controls="basic-navbar-nav"
              label="Toggle navigation"
            >
              <button
                aria-controls="basic-navbar-nav"
                aria-label="Toggle navigation"
                className="navbar-toggler collapsed"
                onClick={[Function]}
                type="button"
              >
                <span
                  className="navbar-toggler-icon"
                />
              </button>
            </NavbarToggle>
            <NavbarCollapse
              id="basic-navbar-nav"
            >
              <Collapse
                appear={false}
                dimension="height"
                getDimensionValue={[Function]}
                id="basic-navbar-nav"
                in={false}
                mountOnEnter={false}
                timeout={300}
                unmountOnExit={false}
              >
                <Transition
                  addEndListener={[Function]}
                  appear={false}
                  aria-expanded={null}
                  enter={true}
                  exit={true}
                  id="basic-navbar-nav"
                  in={false}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntered={[Function]}
                  onEntering={[Function]}
                  onExit={[Function]}
                  onExited={[Function]}
                  onExiting={[Function]}
                  timeout={300}
                  unmountOnExit={false}
                >
                  <div
                    aria-expanded={null}
                    className="navbar-collapse collapse"
                    id="basic-navbar-nav"
                  >
                    <Nav
                      className="mr-auto"
                      fill={false}
                      justify={false}
                    >
                      <ForwardRef
                        as="div"
                        className="mr-auto navbar-nav"
                        onSelect={[Function]}
                      >
                        <div
                          className="mr-auto navbar-nav"
                          onKeyDown={[Function]}
                        >
                          <NavLink
                            as={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "displayName": "SafeAnchor",
                                "render": [Function],
                              }
                            }
                            disabled={false}
                            href="/add-new"
                          >
                            <ForwardRef
                              as={
                                Object {
                                  "$$typeof": Symbol(react.forward_ref),
                                  "displayName": "SafeAnchor",
                                  "render": [Function],
                                }
                              }
                              className="nav-link"
                              disabled={false}
                              href="/add-new"
                            >
                              <SafeAnchor
                                className="nav-link"
                                data-rb-event-key="/add-new"
                                disabled={false}
                                href="/add-new"
                                onClick={[Function]}
                              >
                                <a
                                  className="nav-link"
                                  data-rb-event-key="/add-new"
                                  href="/add-new"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                >
                                  Add New Feedback
                                </a>
                              </SafeAnchor>
                            </ForwardRef>
                          </NavLink>
                        </div>
                      </ForwardRef>
                    </Nav>
                  </div>
                </Transition>
              </Collapse>
            </NavbarCollapse>
          </nav>
        </Navbar>
      </NavBar>
      <Form
        inline={false}
        noValidate={true}
        onSubmit={[Function]}
        validated={false}
      >
        <form
          className=""
          noValidate={true}
          onSubmit={[Function]}
        >
          <FormGroup
            controlId="formBasicTitle"
          >
            <div
              className="form-group"
            >
              <FormLabel
                column={false}
                srOnly={false}
              >
                <label
                  className="form-label"
                  htmlFor="formBasicTitle"
                >
                  Title
                </label>
              </FormLabel>
              <FormControl
                placeholder="Title"
                required={true}
                type="title"
              >
                <input
                  className="form-control"
                  id="formBasicTitle"
                  placeholder="Title"
                  required={true}
                  type="title"
                />
              </FormControl>
              <Feedback
                type="invalid"
              >
                <div
                  className="invalid-feedback"
                >
                  Please Provide Feedback title
                </div>
              </Feedback>
            </div>
          </FormGroup>
          <FormGroup
            controlId="formBasicType"
          >
            <div
              className="form-group"
            >
              <FormLabel
                column={false}
                srOnly={false}
              >
                <label
                  className="form-label"
                  htmlFor="formBasicType"
                >
                  Type
                </label>
              </FormLabel>
              <FormControl
                as="select"
                required={true}
              >
                <select
                  className="form-control"
                  id="formBasicType"
                  required={true}
                >
                  <option
                    value=""
                  >
                    -- Select --
                  </option>
                  <option
                    value="Technical"
                  >
                    Technical
                  </option>
                  <option
                    value="User"
                  >
                    User
                  </option>
                  <option
                    value="Management Related"
                  >
                    Management Related
                  </option>
                  <option
                    value="competitor Related"
                  >
                    competitor Related
                  </option>
                  <option
                    value="Pricing issue"
                  >
                    Pricing issue
                  </option>
                  <option
                    value="Customer Care issue"
                  >
                    Customer Care issue
                  </option>
                </select>
              </FormControl>
              <Feedback
                type="invalid"
              >
                <div
                  className="invalid-feedback"
                >
                  Please Select Feedback type
                </div>
              </Feedback>
            </div>
          </FormGroup>
          <FormGroup
            controlId="formBasicEmail"
          >
            <div
              className="form-group"
            >
              <FormLabel
                column={false}
                srOnly={false}
              >
                <label
                  className="form-label"
                  htmlFor="formBasicEmail"
                >
                  Email address
                </label>
              </FormLabel>
              <FormControl
                onChange={[MockFunction]}
                placeholder="Enter email"
                required={true}
                type="email"
              >
                <input
                  className="form-control"
                  id="formBasicEmail"
                  onChange={[MockFunction]}
                  placeholder="Enter email"
                  required={true}
                  type="email"
                />
              </FormControl>
              <Feedback
                type="invalid"
              >
                <div
                  className="invalid-feedback"
                >
                  Please Enter valid Email
                </div>
              </Feedback>
            </div>
          </FormGroup>
          <FormGroup
            controlId="formBasicDescription"
          >
            <div
              className="form-group"
            >
              <FormLabel
                column={false}
                srOnly={false}
              >
                <label
                  className="form-label"
                  htmlFor="formBasicDescription"
                >
                  Description
                </label>
              </FormLabel>
              <FormControl
                as="textarea"
                required={true}
                rows="3"
              >
                <textarea
                  className="form-control"
                  id="formBasicDescription"
                  required={true}
                  rows="3"
                />
              </FormControl>
              <Feedback
                type="invalid"
              >
                <div
                  className="invalid-feedback"
                >
                  Please Enter Feedback Description
                </div>
              </Feedback>
            </div>
          </FormGroup>
          <Alert
            closeLabel="Close alert"
            show={true}
            transition={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "appear": false,
                  "in": false,
                  "mountOnEnter": false,
                  "timeout": 300,
                  "unmountOnExit": false,
                },
                "displayName": "Fade",
                "render": [Function],
              }
            }
            variant="success"
          >
            <Fade
              appear={false}
              in={true}
              mountOnEnter={false}
              timeout={300}
              unmountOnExit={true}
            >
              <Transition
                addEndListener={[Function]}
                appear={false}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={true}
              >
                <div
                  className="fade alert alert-success show"
                  role="alert"
                >
                  Created Successfully
                </div>
              </Transition>
            </Fade>
          </Alert>
          <Button
            active={false}
            disabled={false}
            type="submit"
            variant="primary"
          >
            <button
              className="btn btn-primary"
              disabled={false}
              type="submit"
            >
              Submit
            </button>
          </Button>
        </form>
      </Form>
    </div>
  </Container>
</NewFeedBack>
`;
